### ISSUE TITLE
StringIndexOutOfBoundException in Lexer.notifyListeners

### ISSUE THREAD

**Issue by scadgek** [antlr/antlr4#2138]
I've stumbled into a bug with handling invalid input by the generated Parser. It is reproduced with, for example, input `x = 'asd`. I'm not sure what the expected result it, probably ignoring invalid `'` character and matching `asd` as ID or throwing an error about invalid character from the parser. Either way, currently it's just a java exception concerning an error when retrieving the offending symbol from the input.
The stacktrace below indicates the class and line where the error happens (tldr: Lexer.java::360).

The following are the grammar and test java file to reproduce the error.
**Test.g4:**
```
grammar Test;

condition
    : (ID | CHAR_STRING) '=' (ID | CHAR_STRING)
    ;

ID: [a-zA-Z]+;

CHAR_STRING: '\'' ~('\'')* '\'';

WS: [ \t\r\n]+ -> skip;

```
**Main.java snippet:**
```
final CodePointCharStream input = CharStreams.fromString( "x = 'q" );
final TestLexer lexer = new TestLexer( input );
final CommonTokenStream tokens = new CommonTokenStream( lexer );
final TestParser parser = new TestParser( tokens );

ParseTreeWalker.DEFAULT.walk( new TestBaseListener(), parser.condition() );
```
And here's the exception thrown:

```
Exception in thread "main" java.lang.StringIndexOutOfBoundsException: String index out of range: 7
	at java.lang.String.checkBounds(String.java:385)
	at java.lang.String.<init>(String.java:462)
	at org.antlr.v4.runtime.CodePointCharStream$CodePoint8BitCharStream.getText(CodePointCharStream.java:160)
	at org.antlr.v4.runtime.Lexer.notifyListeners(Lexer.java:360)
	at org.antlr.v4.runtime.Lexer.nextToken(Lexer.java:144)
	at org.antlr.v4.runtime.BufferedTokenStream.fetch(BufferedTokenStream.java:169)
	at org.antlr.v4.runtime.BufferedTokenStream.sync(BufferedTokenStream.java:152)
	at org.antlr.v4.runtime.BufferedTokenStream.consume(BufferedTokenStream.java:136)
	at org.antlr.v4.runtime.Parser.consume(Parser.java:571)
	at org.antlr.v4.runtime.Parser.match(Parser.java:203)
	at antlr.sample.gen.TestParser.condition(TestParser.java:123)
	at antlr.sample.Main.main(Main.java:20)
```

**Comment by scadgek** [comment]
Oh, looks like it's already fixed in the repository version in commit https://github.com/antlr/antlr4/commit/ac9f75303e215dbc12937eba68089151aef6f874
Will be waiting for the next release, till then I'll have to use the manually built jar.

**Comment by scadgek** [comment]
For anyone waiting for this fix like me: the updated 4.7.1 version is already available on Maven Central: http://mvnrepository.com/artifact/org.antlr/antlr4-runtime/4.7.1

**Comment by scadgek** [comment]
Actually, the same error still occurs for cyrillic letters. To reproduce, in the description above change the test string to "к = 'q".
